README.md
# Ansible Collection - stackhpc.openstack

This collection includes content for interacting with OpenStack clouds.

## Included content

### Roles

- [os_openstacksdk](roles/os_openstacksdk): Install `openstacksdk` Python package


CHANGELOG.rst
====================================
stackhpc.openstack Release Notes
====================================

.. contents:: Topics


v1.0.0
======


plugins/modules/baremetal_deploy_template.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# Copyright (c) 2023 StackHPC Ltd.
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)

# This module has been proposed to the OpenStack collection:
# https://review.opendev.org/c/openstack/ansible-collections-openstack/+/874755
# Keep a copy here until it merges.
# It requires OpenStack SDK 0.53.0.

DOCUMENTATION = r'''
module: baremetal_deploy_template
short_description: Create/Delete Bare Metal deploy template Resources from OpenStack
author: OpenStack Ansible SIG
description:
    - Create, Update and Remove ironic deploy templates from OpenStack.
options:
    extra:
      description:
        - A set of one or more arbitrary metadata key and value pairs.
      type: dict
    id:
      description:
        - ID of the deploy template.
        - Will be auto-generated if not specified.
      type: str
      aliases: ['uuid']
    name:
      description:
        - Name of the deploy template.
        - Must be formatted as a trait name (see API reference).
        - Required when the deploy template is created, after which the
          name or ID may be used.
      type: str
    steps:
      description:
        - List of deploy steps to apply.
        - Required when the deploy template is created.
      type: list
      elements: dict
    state:
      description:
        - Indicates desired state of the resource
      choices: ['present', 'absent']
      default: present
      type: str
extends_documentation_fragment:
- openstack.cloud.openstack
'''

EXAMPLES = r'''
- name: Create Bare Metal deploy template
  openstack.cloud.baremetal_deploy_template:
    cloud: devstack
    state: present
    name: CUSTOM_FOO
    steps:
      - interface: bios
        step: apply_configuration
        args:
          settings:
            - name: LogicalProc
              value: Enabled
        priority: 110
    extra:
      something: extra
  register: result

- name: Delete Bare Metal deploy template
  openstack.cloud.baremetal_deploy_template:
    cloud: devstack
    state: absent
    id: 1a85ebca-22bf-42eb-ad9e-f640789b8098
  register: result

- name: Update Bare Metal deploy template
  openstack.cloud.baremetal_deploy_template:
    cloud: devstack
    state: present
    id: 1a85ebca-22bf-42eb-ad9e-f640789b8098
    extra:
      something: new
'''

RETURN = r'''
template:
    description: A deploy template dictionary, subset of the dictionary keys
                 listed below may be returned, depending on your cloud
                 provider.
    returned: success
    type: dict
    contains:
        created_at:
            description: Bare Metal deploy template created at timestamp.
            returned: success
            type: str
        extra:
            description: A set of one or more arbitrary metadata key and value
                         pairs.
            returned: success
            type: dict
        id:
            description: The UUID for the Baremetal Deploy Template resource.
            returned: success
            type: str
        links:
            description: A list of relative links, including the self and
                         bookmark links.
            returned: success
            type: list
        location:
            description: Cloud location of this resource (cloud, project,
                         region, zone)
            returned: success
            type: dict
        name:
            description: Bare Metal deploy template name.
            returned: success
            type: str
        steps:
            description: A list of deploy steps.
            returned: success
            type: list
            elements: dict
        updated_at:
            description: Bare Metal deploy template updated at timestamp.
            returned: success
            type: str
'''

from ansible_collections.openstack.cloud.plugins.module_utils.openstack import (
    OpenStackModule
)


class BaremetalDeployTemplateModule(OpenStackModule):
    argument_spec = dict(
        extra=dict(type='dict'),
        id=dict(aliases=['uuid']),
        name=dict(),
        steps=dict(type='list', elements='dict'),
        state=dict(default='present', choices=['present', 'absent']),
    )

    module_kwargs = dict(
        required_one_of=[
            ('id', 'name'),
        ],
    )

    def run(self):
        template = self._find_deploy_template()
        state = self.params['state']
        if state == 'present':
            # create or update deploy template

            kwargs = {}
            for k in ['extra', 'id', 'name', 'steps']:
                if self.params[k] is not None:
                    kwargs[k] = self.params[k]

            changed = True
            if not template:
                # create deploy template
                template = self.conn.baremetal.create_deploy_template(**kwargs)
            else:
                # update deploy template
                updates = dict((k, v)
                               for k, v in kwargs.items()
                               if v != template[k])

                if updates:
                    template = \
                        self.conn.baremetal.update_deploy_template(template['id'], **updates)
                else:
                    changed = False

            self.exit_json(changed=changed, template=template.to_dict(computed=False))

        if state == 'absent':
            # remove deploy template
            if not template:
                self.exit_json(changed=False)

            template = self.conn.baremetal.delete_deploy_template(template['id'])
            self.exit_json(changed=True)

    def _find_deploy_template(self):
        id_or_name = self.params['id'] if self.params['id'] else self.params['name']
        try:
            return self.conn.baremetal.get_deploy_template(id_or_name)
        except self.sdk.exceptions.ResourceNotFound:
            return None


def main():
    module = BaremetalDeployTemplateModule()
    module()


if __name__ == "__main__":
    main()


roles/os_ironic_state/library/os_ironic_state.py
#!/usr/bin/python
# coding: utf-8 -*-

# (c) 2015, Hewlett-Packard Development Company, L.P.
# Copyright 2017 StackHPC Ltd.
#
# This module is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This software is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this software.  If not, see <http://www.gnu.org/licenses/>.

DOCUMENTATION = '''
---
module: os_ironic_state
short_description: Set provision state of Bare Metal Resources from OpenStack
author: "Mark Goddard <mark@stackhpc.com>"
extends_documentation_fragment: openstack
description:
    - Set the provision state of OpenStack ironic bare metal nodes.
options:
    provision_state:
      description:
        - Indicates desired provision state of the resource
      choices: ['manage', 'provide']
      default: present
    uuid:
      description:
        - globally unique identifier (UUID) to be given to the resource.
      required: false
      default: None
    ironic_url:
      description:
        - If noauth mode is utilized, this is required to be set to the
          endpoint URL for the Ironic API.  Use with "auth" and "auth_type"
          settings set to None.
      required: false
      default: None
    wait:
      description:
        - A boolean value instructing the module to wait for node
          activation or deactivation to complete before returning.
      required: false
      default: False
      version_added: "2.1"
    timeout:
      description:
        - An integer value representing the number of seconds to
          wait for the node activation or deactivation to complete.
      version_added: "2.1"
    availability_zone:
      description:
        - Ignored. Present for backwards compatibility
      required: false
'''

EXAMPLES = '''
os_ironic_node:
  cloud: "openstack"
  uuid: "d44666e1-35b3-4f6b-acb0-88ab7052da69"
  provision_state: provide
  delegate_to: localhost
'''

from distutils.version import StrictVersion

# Store a list of import errors to report to the user.
IMPORT_ERRORS = []
try:
    import openstack
except Exception as e:
    IMPORT_ERRORS.append(e)

from ansible.module_utils.basic import *
from ansible.module_utils.openstack import *


def _choose_id_value(module):
    if module.params['uuid']:
        return module.params['uuid']
    if module.params['name']:
        return module.params['name']
    return None


def _change_required(current_provision_state, action):
    """Return whether a change to the provision state is required.

    :param current_provision_state: The current provision state of the node.
    :param action: The requested action.
    """
    if action == 'manage':
        if current_provision_state == 'manageable':
            return False
    if action == 'provide':
        if current_provision_state == 'available':
            return False
    return True


def main():
    argument_spec = openstack_full_argument_spec(
        uuid=dict(required=False),
        name=dict(required=False),
        ironic_url=dict(required=False),
        provision_state=dict(required=True,
                             choices=['manage', 'provide']),
        wait=dict(type='bool', required=False, default=False),
        timeout=dict(required=False, type='int', default=1800),
    )
    module_kwargs = openstack_module_kwargs()
    module = AnsibleModule(argument_spec, **module_kwargs)
    # Fail if there were any exceptions when importing modules.
    if IMPORT_ERRORS:
        module.fail_json(msg="Import errors: %s" %
                         ", ".join([repr(e) for e in IMPORT_ERRORS]))

    if (module.params['auth_type'] in [None, 'None'] and
            module.params['ironic_url'] is None):
        module.fail_json(msg="Authentication appears disabled, Please "
                             "define an ironic_url parameter")

    if (module.params['ironic_url'] and
            module.params['auth_type'] in [None, 'None']):
        module.params['auth'] = dict(
            endpoint=module.params['ironic_url']
        )

    node_id = _choose_id_value(module)

    if not node_id:
        module.fail_json(msg="A uuid or name value must be defined "
                             "to use this module.")

    try:
        sdk, cloud = openstack_cloud_from_module(module)
        node = cloud.get_machine(node_id)

        if node is None:
            module.fail_json(msg="node not found")

        uuid = node['uuid']
        changed = False
        wait = module.params['wait']
        timeout = module.params['timeout']
        provision_state = module.params['provision_state']

        if node['provision_state'] in [
                'cleaning',
                'deleting',
                'wait call-back']:
            module.fail_json(msg="Node is in %s state, cannot act upon the "
                                 "request as the node is in a transition "
                                 "state" % node['provision_state'])

        if _change_required(node['provision_state'], provision_state):
            cloud.node_set_provision_state(uuid, provision_state, wait=wait,
                                           timeout=timeout)
            changed = True

        module.exit_json(changed=changed)

    except Exception as e:
        module.fail_json(msg=str(e))


if __name__ == "__main__":
    main()


plugins/modules/baremetal_node_state.py
#!/usr/bin/python
# coding: utf-8 -*-

# (c) 2015, Hewlett-Packard Development Company, L.P.
# Copyright 2017 StackHPC Ltd.
#
# This module is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This software is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this software.  If not, see <http://www.gnu.org/licenses/>.

from __future__ import (absolute_import, division, print_function)
__metaclass__ = type

DOCUMENTATION = '''
---
module: baremetal_node_state
short_description: Set provision state of Bare Metal Resources from OpenStack
author: "Mark Goddard <mark@stackhpc.com>"
extends_documentation_fragment: openstack
description:
    - Set the provision state of OpenStack ironic bare metal nodes.
options:
    provision_state:
      description:
        - Indicates desired provision state of the resource
      choices: ['manage', 'provide']
      default: present
    uuid:
      description:
        - globally unique identifier (UUID) to be given to the resource.
      required: false
      default: None
    ironic_url:
      description:
        - If noauth mode is utilized, this is required to be set to the
          endpoint URL for the Ironic API.  Use with "auth" and "auth_type"
          settings set to None.
      required: false
      default: None
    wait:
      description:
        - A boolean value instructing the module to wait for node
          activation or deactivation to complete before returning.
      required: false
      default: False
      version_added: "2.1"
    timeout:
      description:
        - An integer value representing the number of seconds to
          wait for the node activation or deactivation to complete.
      version_added: "2.1"
    availability_zone:
      description:
        - Ignored. Present for backwards compatibility
      required: false
'''

EXAMPLES = '''
baremeta_node_state:
  cloud: "openstack"
  uuid: "d44666e1-35b3-4f6b-acb0-88ab7052da69"
  provision_state: provide
  delegate_to: localhost
'''

from distutils.version import StrictVersion

# Store a list of import errors to report to the user.
IMPORT_ERRORS = []
try:
    import openstack
except Exception as e:
    IMPORT_ERRORS.append(e)

from ansible.module_utils.basic import *
from ansible.module_utils.openstack import *


def _choose_id_value(module):
    if module.params['uuid']:
        return module.params['uuid']
    if module.params['name']:
        return module.params['name']
    return None


def _change_required(current_provision_state, action):
    """Return whether a change to the provision state is required.

    :param current_provision_state: The current provision state of the node.
    :param action: The requested action.
    """
    if action == 'manage':
        if current_provision_state == 'manageable':
            return False
    if action == 'provide':
        if current_provision_state == 'available':
            return False
    return True


def main():
    argument_spec = openstack_full_argument_spec(
        uuid=dict(required=False),
        name=dict(required=False),
        ironic_url=dict(required=False),
        provision_state=dict(required=True,
                             choices=['manage', 'provide']),
        wait=dict(type='bool', required=False, default=False),
        timeout=dict(required=False, type='int', default=1800),
    )
    module_kwargs = openstack_module_kwargs()
    module = AnsibleModule(argument_spec, **module_kwargs)
    # Fail if there were any exceptions when importing modules.
    if IMPORT_ERRORS:
        module.fail_json(msg="Import errors: %s" %
                         ", ".join([repr(e) for e in IMPORT_ERRORS]))

    if (module.params['auth_type'] in [None, 'None'] and
            module.params['ironic_url'] is None):
        module.fail_json(msg="Authentication appears disabled, Please "
                             "define an ironic_url parameter")

    if (module.params['ironic_url'] and
            module.params['auth_type'] in [None, 'None']):
        module.params['auth'] = dict(
            endpoint=module.params['ironic_url']
        )

    node_id = _choose_id_value(module)

    if not node_id:
        module.fail_json(msg="A uuid or name value must be defined "
                             "to use this module.")

    try:
        sdk, cloud = openstack_cloud_from_module(module)
        node = cloud.get_machine(node_id)

        if node is None:
            module.fail_json(msg="node not found")

        uuid = node['uuid']
        changed = False
        wait = module.params['wait']
        timeout = module.params['timeout']
        provision_state = module.params['provision_state']

        if node['provision_state'] in [
                'cleaning',
                'deleting',
                'wait call-back']:
            module.fail_json(msg="Node is in %s state, cannot act upon the "
                                 "request as the node is in a transition "
                                 "state" % node['provision_state'])

        if _change_required(node['provision_state'], provision_state):
            cloud.node_set_provision_state(uuid, provision_state, wait=wait,
                                           timeout=timeout)
            changed = True

        module.exit_json(changed=changed)

    except Exception as e:
        module.fail_json(msg=str(e))


if __name__ == "__main__":
    main()
